internal/cmd/fbgen/cmd/semantic.go

```diff
@@ -0,0 +1,525 @@
+package cmd
+
+import (
+	"fmt"
+	"go/importer"
+	"go/token"
+	"go/types"
+	"io/ioutil"
+	"log"
+	"os"
+	"strings"
+
+	"github.com/dave/jennifer/jen"
+	"github.com/influxdata/flux/codes"
+	"github.com/influxdata/flux/internal/errors"
+	"github.com/spf13/cobra"
+)
+
+// generateCmd represents the generate command
+var generateCmd = &cobra.Command{
+	Use:   "semantic",
+	Short: "Generate translation code from semantic FlatBuffers to Go semantic graph",
+	RunE:  generateSemantic,
+}
+
+var (
+	flagOutput            string
+	flagContinueOnMissing bool
+)
+
+func init() {
+	rootCmd.AddCommand(generateCmd)
+	generateCmd.Flags().StringVar(&flagOutput, "output", "",
+		"When present, output will be written to the specified file. Otherwise, output is piped to stdout.")
+	generateCmd.Flags().BoolVar(&flagContinueOnMissing, "continue-on-missing", false,
+		"When set, TODO comments will be put in generated code for missing types, instead of terminating with an error.")
+}
+
+const (
+	semPath    = "github.com/influxdata/flux/semantic"
+	fbsemPath  = "github.com/influxdata/flux/semantic/internal/fbsemantic"
+	errorsPath = "github.com/influxdata/flux/internal/errors"
+	codesPath  = "github.com/influxdata/flux/codes"
+)
+
+func generateSemantic(cmd *cobra.Command, args []string) error {
+	if flagOutput != "" {
+		// If there are errors in the generated file because its currently under development,
+		// replace it with something that won't trip up the call to Import() below.
+		if err := ioutil.WriteFile(flagOutput, []byte("package semantic\n"), 0644); err != nil {
+			return err
+		}
+	}
+	pkg, err := importer.ForCompiler(&token.FileSet{}, "source", nil).Import("github.com/influxdata/flux/semantic")
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	f := jen.NewFilePath(semPath)
+	f.Commentf("DO NOT EDIT.  This file was generated by the fbgen command.")
+	f.Line()
+
+	scope := pkg.Scope()
+	for _, n := range scope.Names() {
+		o := scope.Lookup(n)
+		if skipObject(o) {
+			continue
+		}
+
+		st, ok := o.Type().Underlying().(*types.Struct)
+		if !ok {
+			continue
+		}
+		cs := make([]jen.Code, st.NumFields())
+		cs = append(cs, jen.Var().Err().Error())
+		cs = append(cs,
+			jen.If(jen.Id("fb").Op("==").Nil()).Block(
+				jen.Return(jen.Nil()),
+			),
+		)
+		for i := 0; i < st.NumFields(); i++ {
+			c, err := doField(o, st.Field(i))
+			if err != nil {
+				return err
+			}
+			if c != nil {
+				cs = append(cs, c...)
+			}
+		}
```

Comment:
I'm not totally understanding this section, but I'm guessing that this is going through the files and looking for places to generate code based on the function definitions below.